### PR TITLE
ESQL: Fix bug when combining projections

### LINKED ISSUE #107083: DROP of EVAL planning issues
```js
PUT /test/_doc/1?refresh
{
    "k": "a",
    "v": 1
}
```

```js
PUT /test/_doc/2?refresh
{
    "k": "a",
    "v": 2
}
```

```js
POST /_query?error_trace
{
    "query": "FROM test | EVAL k1 = k | STATS avg(v) BY k, k1 | DROP k1"
}
```

```
"reason": "Found 1 problem\nline 1:27: Plan [Aggregate[[k{f}#199, k1{r}#192],[SUM(v{f}#198) AS $$SUM$avg(v)$0, COUNT(v{f}#198) AS $$COUNT$avg(v)$1, k{f}#199]]] optimized incorrectly due to missing references [k1{r}#192]",
```

And  without aggs

```js
POST /_query?error_trace
{
    "query": "FROM test | EVAL k1 = k | STATS BY k, k1 | DROP k1"
}
```

```
"stack_trace": "java.lang.NullPointerException: Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelSet.nameIds()\" because the return value of 
 \"java.util.List.get(int)\" is null
\tat org.elasticsearch.xpack.esql.planner.ExchangeLayout.get(ExchangeLayout.java:36)
\tat org.elasticsearch.xpack.esql.planner.AbstractPhysicalOperationProviders.groupingPhysicalOperation(AbstractPhysicalOperationProviders.java:122)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planAggregation(LocalExecutionPlanner.java:234)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:182)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planProject(LocalExecutionPlanner.java:508)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:196)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planLimit(LocalExecutionPlanner.java:548)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:200)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.planOutput(LocalExecutionPlanner.java:267)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:224)
\tat org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:167)
\tat org.elasticsearch.xpack.esql.plugin.ComputeService.runCompute(ComputeService.java:441)
\tat org.elasticsearch.xpack.esql.plugin.ComputeService.execute(
        ComputeService.java:208)
\tat org.elasticsearch.xpack.esql.plugin.TransportEsqlQueryAction.lambda$innerExecute$4(TransportEsqlQueryAction.java:166)
```

### PR DESCRIPTION
**Pull request by costin**
Recursive aliases (eval x = 1, x1 = x) were not taken into account when
 combining projections causing the target field to be lost (and only the
 immediate intermediate named expression to be used instead which became
 invalid).

Fix #107083

### CODE REVIEW

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 1]
```diff
@@ -366,6 +412,27 @@ public void testCombineProjectionWithAggregation() {
         assertThat(Expressions.names(agg.groupings()), contains("last_name", "first_name"));
     }
 
+    /**
+     * Expects
+     * Limit[1000[INTEGER]]
+     * \_Aggregate[[last_name{f}#23, first_name{f}#20, k{r}#4],[SUM(salary{f}#24) AS s, last_name{f}#23, first_name{f}#20, first_n
+     * ame{f}#20 AS k]]
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:419]: Potentially out of scope, but: couldn't/shouldn't we prune `k` from the groupings, since it's an alias for `first_name`? `k` doesn't seem to be used in the aggregates, either.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 2]
```diff
@@ -431,19 +426,26 @@ private static List<NamedExpression> combineProjections(
             List<? extends NamedExpression> lower
         ) {
 
-            // collect aliases in the lower list
-            AttributeMap<NamedExpression> aliases = new AttributeMap<>();
+            // collect named expressions declaration in the lower list
+            AttributeMap<NamedExpression> namedExpressions = new AttributeMap<>();
+            // while also collecting the alias map for resolving the source (f1 = 1, f2 = f1, etc..)
+            AttributeMap<Expression> aliases = new AttributeMap<>();
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:432]: nit: comments/var names here are a bit hard to follow.

**Review by alex-spies — APPROVED** [review]
I think this looks good, only minor remarks - except one potentially non optimal pruning in an added test, but that shouldn't block this

**Review by astefan — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:419]: See https://github.com/elastic/elasticsearch/issues/107166

**Comment by costin** [comment]
Raised https://github.com/elastic/elasticsearch/pull/107167 for backport
